src/client/dfuse/pil4dfs/aio.c

```diff
@@ -0,0 +1,552 @@
+			D_GOTO(err, rc = EINVAL);
+		}
+	}
+	if (n_op_dfs == 0)
+		goto org;
+
+	if (n_op_dfs != nr) {
+		if (d_compatible_mode)
+			goto org;
+		DS_ERROR(EINVAL, "io_submit() does not support mixed non-dfs and dfs files yet in"
+			 " regular mode");
+		D_GOTO(err, rc = EINVAL);
+	}
+
+	if (!aio_ctx_obj->inited) {
+		rc = create_ev_eq_for_aio(aio_ctx_obj);
+		if (rc)
+			D_GOTO(err, rc);
+	}
+
+	for (i = 0; i < nr; i++) {
+		op         = ios[i]->aio_lio_opcode;
+		fd         = fd_directed[i] - FD_FILE_BASE;
+		idx_aio_eq = aio_ctx_obj->idx_eq;
+
+		D_ALLOC_PTR(ctx_ev);
+		if (ctx_ev == NULL)
+			D_GOTO(err_loop, rc = ENOMEM);
+
+		rc = daos_event_init(&ctx_ev->ev, aio_eq_list[idx_aio_eq].eq, NULL);
+		if (rc) {
+			DL_ERROR(rc, "daos_event_init() failed");
+			D_GOTO(err_loop, rc = daos_der2errno(rc));
+		}
+		ctx_ev->piocb = ios[i];
+		/* EQs are shared by contexts. Need to save ctx when polling EQs. */
+		ctx_ev->ctx = aio_ctx_obj;
+		D_ALLOC_PTR(req_args);
+		if (req_args == NULL)
+			D_GOTO(err_loop, rc = ENOMEM);
+		d_iov_set(&req_args->iov, (void *)ios[i]->u.c.buf, ios[i]->u.c.nbytes);
+		req_args->sgl.sg_nr     = 1;
+		req_args->sgl.sg_iovs   = &req_args->iov;
+
+		if (op == IO_CMD_PREAD) {
+			daos_event_register_comp_cb(&ctx_ev->ev, aio_req_cb, req_args);
+			rc = dfs_read(d_file_list[fd]->dfs_mt->dfs, d_file_list[fd]->file,
+				      &req_args->sgl, ios[i]->u.c.offset, &read_size, &ctx_ev->ev);
+			if (rc) {
+				rc2 = daos_event_fini(&ctx_ev->ev);
+				if (rc2)
+					DL_ERROR(rc2, "daos_event_fini() failed");
+				D_GOTO(err_loop, rc);
+			}
+		}
+		if (op == IO_CMD_PWRITE) {
+			daos_event_register_comp_cb(&ctx_ev->ev, aio_req_cb, req_args);
+			rc = dfs_write(d_file_list[fd]->dfs_mt->dfs, d_file_list[fd]->file,
+				       &req_args->sgl, ios[i]->u.c.offset, &ctx_ev->ev);
+			if (rc) {
+				rc2 = daos_event_fini(&ctx_ev->ev);
+				if (rc2)
+					DL_ERROR(rc2, "daos_event_fini() failed");
+				D_GOTO(err_loop, rc);
+			}
+		}
+		atomic_fetch_add_relaxed(&aio_ctx_obj->num_op_submitted, 1);
+		atomic_fetch_add_relaxed(&aio_eq_list[idx_aio_eq].n_op_queued, 1);
+	}
+
+	D_FREE(fd_directed);
+	return nr;
+
+org:
+	D_FREE(fd_directed);
+	return next_io_submit(ctx_real, nr, ios);
+
+err:
+	D_FREE(fd_directed);
+	return (-rc);
+
+err_loop:
+	D_FREE(ctx_ev);
+	D_FREE(fd_directed);
+
+	return i ? i : (-rc);
+}
+
+int
+io_cancel(io_context_t ctx, struct iocb *iocb, struct io_event *evt)
+{
+	d_aio_ctx_t *aio_ctx_obj = (d_aio_ctx_t *)ctx;
+	io_context_t ctx_real    = aio_ctx_obj->ctx;
+
+	if (next_io_cancel == NULL) {
+		next_io_cancel = dlsym(RTLD_NEXT, "io_cancel");
+		D_ASSERT(next_io_cancel != NULL);
+	}
+	if (!d_hook_enabled)
+		return next_io_cancel(ctx_real, iocb, evt);
+
+	if (iocb->aio_fildes < FD_FILE_BASE)
+		return next_io_cancel(ctx_real, iocb, evt);
+
+	/* daos_event_abort() may be used to implement this feature later. */
+	DS_ERROR(ENOSYS, "io_cancel() for DFS is not implemented");
+	return (-ENOSYS);
```

Comment:
From the man page:

RETURN VALUES
       0 is returned on success, otherwise returns errno.

This should be return ENOSYS, no?

Reply:
Please see the example code in kernel for io_submit(). Negative error is used.
https://github.com/torvalds/linux/blob/master/fs/aio.c#L2096

Reply:
For io_submit, yes ... but the man page says to use positive error for io_cancel. anyways ....

Reply:
https://github.com/torvalds/linux/blob/master/fs/aio.c#L2193
io_cancel also uses negative errno in kernel.